pkg/v1/utils/volumes.go

```diff
@@ -0,0 +1,152 @@
+	parts := strings.Split(strings.TrimPrefix(k, volumeKeyPrefix), "/")
+	return parts[len(parts)-1]
+}
+
+// podUID gets pod UID from the pod log path.
+func podUID(s *specs.Spec) (string, error) {
+	sandboxLogDir := s.Annotations[annotations.SandboxLogDir]
+	if sandboxLogDir == "" {
+		return "", errors.New("no sandbox log path annotation")
+	}
+	fields := strings.Split(filepath.Base(sandboxLogDir), "_")
+	switch len(fields) {
+	case 1: // This is the old CRI logging path
+		return fields[0], nil
+	case 3: // This is the new CRI logging path
+		return fields[2], nil
+	}
+	return "", errors.Errorf("unexpected sandbox log path %q", sandboxLogDir)
+}
+
+// isVolumeKey checks whether an annotation key is for volume.
+func isVolumeKey(k string) bool {
+	return strings.HasPrefix(k, volumeKeyPrefix)
+}
+
+// volumeSourceKey constructs the annotation key for volume source.
+func volumeSourceKey(volume string) string {
+	return volumeKeyPrefix + volume + "/source"
+}
+
+// volumePath searches the volume path in the kubelet pod directory.
+func volumePath(volume, uid string) (string, error) {
+	// TODO: Support subpath when gvisor supports pod volume bind mount.
+	volumeSearchPath := fmt.Sprintf("/var/lib/kubelet/pods/%s/volumes/*/%s", uid, volume)
+	dirs, err := filepath.Glob(volumeSearchPath)
+	if err != nil {
+		return "", err
+	}
+	if len(dirs) != 1 {
+		return "", errors.Errorf("unexpected matched volume list %v", dirs)
+	}
+	return dirs[0], nil
+}
+
+// isVolumePath checks whether a string is the volume path.
+func isVolumePath(volume, path string) (bool, error) {
+	// TODO: Support subpath when gvisor supports pod volume bind mount.
+	volumeSearchPath := fmt.Sprintf("/var/lib/kubelet/pods/*/volumes/*/%s", volume)
+	return filepath.Match(volumeSearchPath, path)
+}
+
+// UpdateVolumeAnnotations add necessary OCI annotations for gvisor
+// volume optimization.
+func UpdateVolumeAnnotations(bundle string, s *specs.Spec) error {
+	var (
+		uid string
+		err error
+	)
+	if IsSandbox(s) {
+		uid, err = podUID(s)
+		if err != nil {
+			// Skip if we can't get pod UID, because this doesn't work
+			// for containerd 1.1.
+			logrus.WithError(err).Error("Can't get pod uid")
+			return nil
+		}
+	}
+	done := make(map[string]struct{})
+	for k, v := range s.Annotations {
+		if !isVolumeKey(k) {
+			continue
+		}
+		if volumeFieldName(k) != "type" {
+			continue
+		}
+		if v != "tmpfs" {
+			// Only tmpfs is supported now.
+			continue
+		}
+		volume := volumeName(k)
+		if _, ok := done[volume]; ok {
+			continue
+		}
+		if uid != "" {
+			// This is a sandbox
+			path, err := volumePath(volume, uid)
+			if err != nil {
+				return errors.Wrapf(err, "get volume path for %q", volume)
+			}
+			s.Annotations[volumeSourceKey(volume)] = path
+		} else {
+			// This is a container
+			for i := range s.Mounts {
+				if yes, _ := isVolumePath(volume, s.Mounts[i].Source); yes {
+					s.Mounts[i].Type = "tmpfs"
+				}
+			}
+		}
+		done[volume] = struct{}{}
```

Comment:
Why do you need to worry about duplicate volume annotations?

Reply:
Initially I didn't only handle the `type` key, so we need to skip other annotations once one is handled. But now we only handle the `type` key now, but still need to track whether the spec is updated or not, so I left `done` there.

I could change it to a single boolean if you prefer. :)

Reply:
Changed to an `updated` boolean.